Implement Lagrange kernel constraints

plafer opened the pull request:
Closes: #240

Builds on: #245

Done:
- failing test with `Air` which uses a Lagrange kernel auxiliary column
- verifier 

Left to do:
- [x] Prover populates `OodFrame` with Lagrange kernel "frame"
- [x] Prover evaluates Lagrange kernel transition and boundary constraints
- [x] Cleanup (including revisiting current API)

plafer commented on winterfell/src/tests.rs:
```diff
@@ -0,0 +1,206 @@
+    matrix::ColMatrix,
+};
+
+#[cfg(not(feature = "std"))]
+use alloc::{vec, vec::Vec};
```
Note: I wasn't able to make the imports work when using `winter-utils` as a `dev-dependency`, so I opted for this

plafer commented:
@irakliyk @Al-Kindi-0 this is ready for review!

The PR has a bigger diff than I originally anticipated. The main reason is that many existing abstractions did not apply to the Lagrange kernel column. For example, I had to create a `LagrangeKernelEvaluationFrame`, `LagrangeKernelBoundaryConstraint`, and `LagrangeKernelTransitionConstraints`, since the existing `EvaluationFrame`, `BoundaryConstraints` and `TransitionConstraints` made assumptions that didn't apply to the Lagrange kernel column. It would be nice to clean this up, but probably in another PR?

Related: #80

Al-Kindi-0 commented on air/src/air/transition/frame.rs:
```diff
@@ -77,3 +79,64 @@ impl<E: FieldElement> EvaluationFrame<E> {
+///
+/// The Lagrange kernel's evaluation frame is different from [`EvaluationFrame`]. Specifically,
+/// - it only contains evaluations from the Lagrange kernel column
+///   (compared to all rows in the case of [`EvaluationFrame`])
```
should this be "all columns"?

plafer replied:
Yup - fixed

Al-Kindi-0 commented on air/src/air/transition/frame.rs:
```diff
@@ -77,3 +79,64 @@ impl<E: FieldElement> EvaluationFrame<E> {
+/// - it only contains evaluations from the Lagrange kernel column
+///   (compared to all rows in the case of [`EvaluationFrame`])
+/// - The column is evaluated at points `z`, `gz`, `g^2 z`, ..., `g^(2^(v-1)) z`,
+///   where `z` is an arbitrary point, and `g` is the trace domain generator
```
We could just say that z is the OOD point to be more specific

Al-Kindi-0 replied:
I see, it is used for both situations. Then I would probably use x instead of z in order to avoid confusions as z is, implicitly, used throughout the codebase to denote the OOD point.

Al-Kindi-0 commented on air/src/air/transition/frame.rs:
```diff
@@ -77,3 +79,64 @@ impl<E: FieldElement> EvaluationFrame<E> {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Constructs a Lagrange kernel evaluation frame from the raw column polynomial evaluations
```
minor nit: evaluations -> evaluations.

Al-Kindi-0 commented on air/src/air/transition/frame.rs:
```diff
@@ -77,3 +79,64 @@ impl<E: FieldElement> EvaluationFrame<E> {
+        Self { frame }
+    }
+
+    /// Constructs the frame from the Lagrange kernel trace column polynomial coefficients, and an evaluation point.
```
minor nit: "..., and an evaluation" -> "... for an evaluation point"

Al-Kindi-0 commented on air/src/air/transition/frame.rs:
```diff
@@ -77,3 +79,64 @@ impl<E: FieldElement> EvaluationFrame<E> {
+    // ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a reference to the inner frame
```
minor nit: frame -> frame.

Al-Kindi-0 commented on air/src/air/transition/mod.rs:
```diff
@@ -171,3 +173,64 @@ impl<E: FieldElement> TransitionConstraints<E> {
+/// Represents the transition constraints for the Lagrange kernel column, as well as the random
+/// coefficients used to linearly combine all the constraints.
+///
+/// There are `log(trace_len)` constraints, each with their own divisor, as described in
```
should it be "its" instead of "their"? Not sure to be honest

plafer replied:
According to ChatGPT, it should be "its"

>The correct version is: "There are 42 constraints, each with its own divisor."
>
>In this sentence, "each" is referring to each constraint individually, so "its" is the appropriate possessive pronoun to use, as it agrees with the singular noun "constraint."

Fixed.

Al-Kindi-0 commented on air/src/proof/ood_frame.rs:
```diff
@@ -128,6 +169,10 @@ impl Serializable for OodFrame {
         target.write_u16(self.trace_states.len() as u16);
         target.write_bytes(&self.trace_states);
 
+        // write lagrange kernel column trace rows
```
nit: Lagrange

Al-Kindi-0 commented on air/src/proof/ood_frame.rs:
```diff
@@ -149,13 +194,68 @@ impl Deserializable for OodFrame {
         let num_trace_state_bytes = source.read_u16()? as usize;
         let trace_states = source.read_vec(num_trace_state_bytes)?;
 
+        // read lagrange kernel column trace rows
```
nit: Lagrange

Al-Kindi-0 commented on air/src/proof/ood_frame.rs:
```diff
@@ -149,13 +194,68 @@ impl Deserializable for OodFrame {
+// OOD FRAME TRACE STATES
+// ================================================================================================
+
+/// Stores the trace evaluations evaluated at `z` and `gz`, where `z` is a random Field element. If
```
nit: we should probably say: "... trace evaluations at ..."

plafer replied:
Agreed, fixed

Al-Kindi-0 commented on prover/src/constraints/evaluator/default.rs:
```diff
@@ -107,8 +110,26 @@ where
-        evaluation_table.combine()
+        // combine all evaluations into a single column
+        let combined_evaluations = if self.air.trace_info().has_lagrange_kernel_aux_column() {
+            // if present, linearly combine the lagrange kernel evaluations too
```
nit: Lagrange

Al-Kindi-0 commented on prover/src/constraints/evaluator/default.rs:
```diff
@@ -257,6 +313,56 @@ where
+            let domain_point = domain.get_ce_x_at(step);
+
+            let transition_constraints_combined_evals = {
+                let mut transition_evals = E::zeroed_vector(log2(domain.trace_length()) as usize);
```
could we move this outside the loop? would it even matter performance wise?

plafer replied:
This is a tiny buffer, so it wouldn't affect performance. I kept the same pattern as `Air::evaluate_transition()` - i.e. zero out the buffer before passing it in the method. Technically, in both cases, zeroing out shouldn't matter because the method *should* touch every array index. IMO, whether we zero out the buffer or not, I mainly care that we be consistent.

Also, this is one of the reasons why I typically stay away from out parameters - this sort of ambiguity tends to create bugs.

Al-Kindi-0 commented on prover/src/lib.rs:
```diff
@@ -382,7 +388,7 @@ pub trait Prover {
             // evaluate trace and constraint polynomials at the OOD point z, and send the results to
             // the verifier. the trace polynomials are actually evaluated over two points: z and
-            // z * g, where g is the generator of the trace domain.
+            // z * g, where g is the generator of the trace domain. Additionally, if the Lagrange kernel auxiliary column is present, we also evaluate that column over the points: z, z * g, z * g^2, z * g^4, ..., z * g^(2^(v-1)), where v = log(trace_len).
```
nit: could we fix the width of the comment?

Al-Kindi-0 reviewed (commented):
Looks great! Thank you @plafer !
Once we agree on #257 , we should come back and revisit some portions of this PR.

irakliyk reviewed (commented):
Thank you! Looks good! This is not a full review yet - but I wanted to leave some preliminary comments.

One general observation: I wonder if it is possible to make Lagrange constraint evaluation a bit more "self-contained". For example, it may be possible to return something like this in the `Air` trait:
```Rust
trait Air {
    ...

    fn get_lagrange_kernel_evaluator(&self) -> Option<LagrangeKernelEvaluator>

    ...
}
```
This `LagrangeKernelEvaluator` would be responsible for evaluating the constraints for a given `LagrangeKernelEvaluationFrame`.

I haven't digested the code enough to say for sure if this is a good idea or not though.

plafer commented on air/src/air/divisor.rs:
```diff
@@ -54,10 +54,6 @@ impl<B: StarkField> ConstraintDivisor<B> {
     /// The above divisor specifies that transition constraints must hold on all steps of the
     /// execution trace except for the last $k$ steps.
     pub fn from_transition(trace_length: usize, num_exemptions: usize) -> Self {
-        assert!(
-            num_exemptions > 0,
-            "invalid number of transition exemptions: must be greater than zero"
-        );
```
The Lagrange kernel transition constraints don't have any exemptions ([here](https://github.com/plafer/winterfell/blob/8544e17037c3d4c2391d4cded3693e161180aa48/air/src/air/transition/mod.rs#L206))

plafer commented on air/src/air/trace_info.rs:
```diff
@@ -28,6 +28,7 @@ pub struct TraceInfo {
     main_segment_width: usize,
     aux_segment_widths: [usize; NUM_AUX_SEGMENTS],
     aux_segment_rands: [usize; NUM_AUX_SEGMENTS],
+    lagrange_kernel_aux_column_idx: Option<usize>,
```
I agree that it doesn't belong in `TraceInfo` - I probably put it there initially to solve some immediate problems.

However, I moved it to `AirContext`, since the Lagrange column index is an Air parameter. Otherwise `AirContext` is incomplete IMO: it holds main/aux transition degrees, the assertions, and trace information, but nothing about the Lagrange column.

Do you agree? I could also add a method `Air::lagrange_kernel_aux_column_idx()` that forwards to the internal context?

plafer commented on prover/src/trace/trace_lde/default/mod.rs:
```diff
@@ -8,7 +8,9 @@ use super::{
 use crate::{RowMatrix, DEFAULT_SEGMENT_WIDTH};
+use air::LagrangeKernelEvaluationFrame;
 use crypto::MerkleTree;
+use math::log2;
```
I removed the `log2()` function in math and switched everything to use `ilog2()`

plafer commented on air/src/air/mod.rs:
```diff
@@ -285,6 +291,69 @@ pub trait Air: Send + Sync {
 
+    /// Evaluates the Lagrange kernel transition constraints over the specified Lagrange kernel
+    /// evaluation frame and stores them in `result`.
+    fn evaluate_lagrange_kernel_aux_transition<F, E>(
```
I agree. I called it `LagrangeKernelTransitionConstraints::evaluate_and_combine()`, since since `evaluate()` is used across repo to mean just evaluating the numerator

irakliyk commented on air/src/proof/ood_frame.rs:
```diff
@@ -9,18 +9,25 @@ use utils::{
+
+/// Represents an [`OodFrame`] where the trace and constraint evaluations have been parsed out.
+pub struct ParsedOodFrame<E> {
+    pub ood_trace_evaluations: Vec<E>,
+    pub ood_lagrange_kernel_trace_evaluations: Option<Vec<E>>,
+    pub ood_constraint_evaluations: Vec<E>,
```
minor nit: do we need the `ood` prefix here?

plafer replied:
Removed

irakliyk commented on air/src/proof/ood_frame.rs:
```diff
@@ -149,13 +194,68 @@ impl Deserializable for OodFrame {
+/// the Air contains a Lagrange kernel auxiliary column, then that column interpolated polynomial
+/// will be evaluated at `z`, `gz`, `g^2 z`, ... `g^(2^(v-1)) z`, where `v == log(trace_len)`, and
+/// stored in `lagrange_kernel_frame`.
+pub struct OodFrameTraceStates<E: FieldElement> {
```
Is this struct still used somewhere?

plafer replied:
It is used [here](https://github.com/plafer/winterfell/blob/a3377d936b32b196e4a59a87ad39dc27cedffcb8/prover/src/trace/poly_table.rs#L80), [here](https://github.com/plafer/winterfell/blob/a3377d936b32b196e4a59a87ad39dc27cedffcb8/air/src/proof/ood_frame.rs#L57) and [here](https://github.com/plafer/winterfell/blob/a3377d936b32b196e4a59a87ad39dc27cedffcb8/prover/src/composer/mod.rs#L66)

plafer commented on air/src/air/transition/frame.rs:
```diff
@@ -77,3 +79,65 @@ impl<E: FieldElement> EvaluationFrame<E> {
+
+        // push `c(gz)`, `c(z * g^2)`, `c(z * g^4)`, ..., `c(z * g^(2^(v-1)))`
+        for i in 0..log_trace_len {
+            let x = g.exp_vartime(2_u32.pow(i).into()) * z;
```
Actually we're increasing the power nonlinearly. The powers are: 1, 2, 4, 8, 16, ... . We could try to optimize this a bit, but we're only doing this `log(trace_len)` times, i.e. always a relatively few number of times. I think the added complexity wouldn't be worth the marginal performance improvement?

irakliyk replied:
Since this is used only for the OOD frame, using `exp_vartime()` is probably fine - but also, I don't think this would add much complexity.

If the powers are $g$, $g^2$, $g^4$, $g^8$ etc. wouldn't this be just:
```Rust
let mut g_exp = g;
for i in 0..log_trace_len {
    let x = g_exp * z;
    let lagrange_poly_at_x = polynom::eval(lagrange_kernel_col_poly, x);

    frame.push(lagrange_poly_at_x);
    g_exp *= g_exp;
}
```

plafer replied:
bleh, you're right. I changed it to use this code snippet.

plafer commented on air/src/proof/ood_frame.rs:
```diff
@@ -38,29 +46,48 @@ impl OodFrame {
     /// # Panics
     /// Panics if evaluation frame has already been set.
-    pub fn set_trace_states<E: FieldElement>(&mut self, trace_states: &[Vec<E>]) -> Vec<E> {
+    pub fn set_trace_states<E: FieldElement>(
+        &mut self,
+        trace_states: &OodFrameTraceStates<E>,
+    ) -> Vec<E> {
```
I actually like that an Ood frame contains both the "regular" current/next frames, as well as the Lagrange kernel frame. I think of it as one unit; all the data necessary to check the out-of-domain point in one place. I'm curious why you were thinking about separating them though.

> (with this, we may be able to avoid the need for `OodFrameTraceStates` struct)

How do you see we get rid of it? We'll still need to store the current & next frames.

But I also have as mental note to reduce the number of "ood frame" structs, but I was leaving it for another PR :).

plafer commented:
@irakliyk ready for another review! I still have the `read_lagrange_kernel_frame_into()` comment to address - which I agree with and will implement first thing tomorrow.

irakliyk commented on air/src/air/mod.rs:
```diff
@@ -285,6 +291,24 @@ pub trait Air: Send + Sync {
     // PROVIDED METHODS
     // --------------------------------------------------------------------------------------------
 
+    /// Returns the random elements to use in the lagrange kernel.
+    ///
+    /// The return slice stores the least significant random element first. For example, for a trace
+    /// length of 8, there will be 3 random elements, such that
+    /// - return[0] = r0
+    /// - return[1] = r1
+    /// - return[2] = r2
+    /// where `return` is the returned slice, and r = (r2, r1, r0).
+    fn lagrange_kernel_rand_elements<'r, F, E>(&self, aux_rand_elements: &'r [E]) -> &'r [E]
```
I'm not sure we need this function just yet. I think for now we should assume that $r$-values are exogenous to the system. In the next PR, we can figure out where these should come from.

plafer replied:
removed

irakliyk commented on air/src/air/mod.rs:
```diff
@@ -357,13 +381,22 @@ pub trait Air: Send + Sync {
     fn get_boundary_constraints<E: FieldElement<BaseField = Self::BaseField>>(
         &self,
         aux_rand_elements: &AuxTraceRandElements<E>,
-        composition_coefficients: &[E],
+        boundary_composition_coefficients: &[E],
+        lagrange_kernel_boundary_coefficient: Option<E>,
     ) -> BoundaryConstraints<E> {
```
Do we need to integrate Lagrange kernel boundary constraints into other boundary constraints? In my initial thinking, we'd treat all Lagrange constraints separately. Specifically, instead of having 2 types of constraints as we currently do, we'd have:

1. Transition constraints (same as now).
2. Boundary constraints (same as now).
3. Lagrange kernel constraints - this would include boundary and transition constraints relevant to the Lagrange column.

plafer replied:
Similar to other comments, I think we should merge them with `TransitionConstraints` and `BoundaryConstraints` (https://github.com/facebook/winterfell/issues/261)

irakliyk replied:
I think trying to merge them into current `TransitionConstraints` and `BoundaryConstraints` will result in more complexity than required. It is also not clear to me that the strategies we use for evaluating (and optimizing) transition/boundary constraints will be applicable to Lagrange kernel constraints. So, as mention in the previous comment, my preference would be to keep Lagrange kernel constraints as encapsulated as possible and not treat them as special cases of boundary/transition constrains.

Of course, if later on we decided that it is a better approach, we can always refactor the codebase - but for now, I think it would be good to implement end-to-end handling of Lagrange column separately.

Concretely, we could add a method to the `Air` trait which would look something like this:
```Rust
fn get_lagrange_kernel_constraints<E: FieldElement<BaseField = Self::BaseField>>(
    &self,
    rand_elements: &Vec<E>, // these would be the r-values
    composition_coefficients: &LagrangeConstraintsCompositionCoefficients<E>,
) -> Option<LagrangeKernelConstraints<E>>
```
This method would return `None` if there is no Lagrange kernel column in the trace, and `LagrangeKernelConstraints` otherwise. `LagrangeKernelConstraints` would have something like `evaluate_at()` method, which would evaluate all Lagrange kernel constraints (boundary and transition) for a given frame. (it could also have other methods which would help the prover to perform evaluations in a more optimal way).

plafer replied:
I had separated them already; I had forgotten to update the `BoundaryConstraints` structs (the Lagrange constraints part was not actually used anymore). Now fixed.

The `LagrangeKernelConstraints` part of the comment still needs to be addressed though.

irakliyk commented on verifier/src/channel.rs:
```diff
@@ -407,11 +433,18 @@ impl<E: FieldElement> TraceOodFrame<E> {
     }
+
+    /// Returns the Lagrange kernel evaluation frame, if any.
+    pub fn lagrange_kernel_column_frame(&self) -> Option<&LagrangeKernelEvaluationFrame<E>> {
```
minor nit: `lagrange_kernel_frame()` is probably sufficient (i.e., no need for "column").

irakliyk commented on prover/src/constraints/evaluator/default.rs:
```diff
@@ -131,18 +152,44 @@ where
 
+        let lagrange_kernel_transition_constraints =
+            air.context().lagrange_kernel_aux_column_idx().map(|_| {
+                LagrangeKernelTransitionConstraints::new(
+                    air.context(),
+                    composition_coefficients.lagrange_kernel_transition,
+                )
+            });
+
         // build periodic value table
         let periodic_values = PeriodicValueTable::new(air);
 
         // build boundary constraint groups; these will be used to evaluate and compose boundary
         // constraint evaluations.
-        let boundary_constraints =
-            BoundaryConstraints::new(air, &aux_rand_elements, &composition_coefficients.boundary);
+        let boundary_constraints = BoundaryConstraints::new(
+            air,
+            &aux_rand_elements,
+            &composition_coefficients.boundary,
+            composition_coefficients.lagrange_kernel_boundary,
+        );
+
+        let lagrange_kernel_boundary_constraint = composition_coefficients
+            .lagrange_kernel_boundary
+            .map(|lagrange_kernel_boundary_coefficient| {
+                let lagrange_kernel_aux_rand_elements =
+                    air.lagrange_kernel_rand_elements(aux_rand_elements.get_segment_elements(0));
+
+                LagrangeKernelBoundaryConstraint::new(
+                    lagrange_kernel_boundary_coefficient,
+                    lagrange_kernel_aux_rand_elements,
+                )
+            });
```
Instead of separating Lagrange boundary and transition constraints here, I wonder if we should treat them as a single unit - e.g., `LagrangeKernelConstraints` instantiated something like:
```Rust
let lagrange_kernel_constraints =
 air.context().lagrange_kernel_aux_column_idx().map(|_| {
     LagrangeKernelConstraints::new(
         air.context(),
         lagrange_rand_elements, // the r-values
         composition_coefficients.lagrange,
     )
 });
```
And then we split into evaluating Lagrange boundary and transition constraints internally within this struct.

I guess the main question is how to get the $r$-values here.

plafer replied:
I think we should instead generalize the existing `TransitionConstraints` and `BoundaryConstraints`, as detailed in https://github.com/facebook/winterfell/issues/261

irakliyk commented on prover/src/constraints/evaluator/default.rs:
```diff
@@ -233,8 +280,19 @@ where
-            evaluations[0] +=
-                self.evaluate_aux_transition(&main_frame, &aux_frame, step, &mut ta_evaluations);
+
+            // Make sure to only evaluate the aux transition if the user actually defined one
+            if !trace_aux_segment_has_only_lagrange_kernel_column(
+                self.air.context().lagrange_kernel_aux_column_idx(),
+                self.air.trace_info(),
+            ) {
+                evaluations[0] += self.evaluate_aux_transition(
+                    &main_frame,
+                    &aux_frame,
+                    step,
+                    &mut ta_evaluations,
+                );
+            }
```
If there are no other aux constraints - shouldn't we just call `evaluate_fragment_main()` instead of calling `evaluate_fragment_full()` and then having this check?

plafer replied:
I since removed `trace_aux_segment_has_only_lagrange_kernel_column()`, since we don't want to support `Air` that only have a Lagrange kernel column (from previous comment). So the previous behavior where we call `evaluate_aux_transition()` unconditionally is now restored.

irakliyk reviewed (commented):
Thank you! Looks good! Still not a full review - but I left some more comments inline.

plafer commented on air/src/air/coefficients.rs:
```diff
@@ -65,7 +65,9 @@ impl<E: FieldElement> Default for AuxTraceRandElements<E> {
 #[derive(Debug, Clone)]
 pub struct ConstraintCompositionCoefficients<E: FieldElement> {
     pub transition: Vec<E>,
+    pub lagrange_kernel_transition: Vec<E>,
     pub boundary: Vec<E>,
+    pub lagrange_kernel_boundary: Option<E>,
 }
```
What benefits do you see in doing this? I typically don't like this style, since the type becomes much less descriptive, and more bug-prone. To understand what the `lagrange` field is, I have to go see how it is populated in the code. And I need to keep in my head that `lagrange[0]` is different from all other `lagrange[i]`, and lookout for bugs e.g. that loop over all indices and apply a "transition constraint operation".

plafer replied:
Replaced both fields with a single `LagrangeConstraintsCompositionCoefficients` one

plafer commented on prover/src/trace/trace_lde/mod.rs:
```diff
@@ -56,6 +56,13 @@ pub trait TraceLde<E: FieldElement>: Sync {
+    /// Note that unlike [`EvaluationFrame`], the Lagrange kernel frame includes only the Lagrange
+    /// kernel column (as opposed to all columns).
+    fn get_lagrange_kernel_column_frame(&self, lde_step: usize)
+        -> LagrangeKernelEvaluationFrame<E>;
```
done

plafer commented on air/src/air/boundary/constraint.rs:
```diff
@@ -142,3 +162,54 @@ where
+        let denominator = x - E::ONE;
+
+        numerator / denominator
+    }
```
done

plafer commented on air/src/air/context.rs:
```diff
@@ -119,6 +132,14 @@ impl<B: StarkField> AirContext<B> {
+            assert!(lagrange_kernel_aux_column_idx < trace_info.get_aux_segment_width(0), "Lagrange kernel column index out of bounds: index={}, but only {} columns in segment", lagrange_kernel_aux_column_idx, trace_info.get_aux_segment_width(0));
+
+            let min_aux_segment_rands = trace_info.length().ilog2();
+            assert!(trace_info.get_aux_segment_rand_elements(0) >= min_aux_segment_rands as usize, "Lagrange kernel column requires log(trace_length) random elements. Got: {}, but need at least {}", trace_info.get_aux_segment_rand_elements(0), min_aux_segment_rands);
```
Removed

plafer commented:
Added a benchmark that runs the prover on a trace with a Lagrange kernel column, with trace lengths `2^14`, `2^15`, `2^16`. On my Macbook Pro M2, the results were:

- trace length $2^{14}$: [426.12 ms 426.70 ms 427.21 ms]
- trace length $2^{15}$: [906.85 ms 908.47 ms 910.41 ms]
- trace length $2^{16}$: [1.9335 s 1.9348 s 1.9363 s]

plafer commented on air/src/air/transition/mod.rs:
```diff
@@ -171,3 +172,88 @@ impl<E: FieldElement> TransitionConstraints<E> {
+                let z = divisor.evaluate_at(x);
+
+                acc + (coef.mul_base(const_eval) / z.into())
+            })
```
Fixed. I ended up encapsulating the logic in a `LagrangeConstraintsBatchEvaluator` that lives next to `DefaultConstraintEvaluator`. The idea is that this is an algorithm that is specific to the prover (hence lives in the prover crate), and runs on the CPU (hence lives next to `DefaultConstraintEvaluator`).

plafer commented:
@irakliyk this is ready for review!

Benchmarks after optimization:

- trace length $2^{14}$: [81.791 ms 81.859 ms 81.918 ms]
- trace length $2^{15}$: [164.98 ms 165.21 ms 165.57 ms]
- trace length $2^{16}$: [352.78 ms 353.17 ms 353.56 ms]

A 5x speedup!

irakliyk commented on air/src/air/transition/frame.rs:
```diff
@@ -77,3 +79,80 @@ impl<E: FieldElement> EvaluationFrame<E> {
+///   (compared to all columns in the case of [`EvaluationFrame`])
+/// - The column is evaluated at points `x`, `gx`, `g^2 x`, ..., `g^(2^(v-1)) x`,
+///   where `x` is an arbitrary point, and `g` is the trace domain generator
+#[derive(Debug, Clone)]
+pub struct LagrangeKernelEvaluationFrame<E: FieldElement> {
+    frame: Vec<E>,
+}
```
I would move this to `src/air/lagrange/frame.rs` module.

irakliyk commented on air/src/air/transition/mod.rs:
```diff
@@ -171,3 +172,137 @@ impl<E: FieldElement> TransitionConstraints<E> {
+///
+/// There are `log(trace_len)` constraints, each with its own divisor, as described in
+/// [this issue](https://github.com/facebook/winterfell/issues/240).
+pub struct LagrangeKernelTransitionConstraints<E: FieldElement> {
+    lagrange_constraint_coefficients: Vec<E>,
+    divisors: Vec<ConstraintDivisor<E::BaseField>>,
+}
```
I would move this to `src/air/lagrange/transition.rs` module.

Also, I'm not sure we need to use `ConstraintDivisor` structs here. The divisors for Lagrange kernel constraints are pretty simple and I think it might be simpler (and probably more efficient) to handle them in a specialized way here.

irakliyk commented on prover/src/constraints/evaluator/default.rs:
```diff
@@ -130,19 +150,29 @@ where
+            composition_coefficients.lagrange.boundary,
+        );
+
+        let lagrange_constraints_evaluator = LagrangeConstraintsBatchEvaluator::new(
+            air,
+            aux_rand_elements.clone(),
+            composition_coefficients,
+        );
```
I think using `aux_rand_elements` here is fine for now - but my current understanding is that the $r$-values for the Lagrange kernel column will be separate from these. But this is something we'll address in the next PR.

irakliyk commented on prover/src/constraints/evaluator/default.rs:
```diff
@@ -39,6 +40,7 @@ pub struct DefaultConstraintEvaluator<'a, A: Air, E: FieldElement<BaseField = A:
     air: &'a A,
     boundary_constraints: BoundaryConstraints<E>,
     transition_constraints: TransitionConstraints<E>,
+    lagrange_constraints_evaluator: LagrangeConstraintsBatchEvaluator<E>,
```
I would make this an optional field - this way, we won't have to handle optionality inside `LagrangeConstraintsBatchEvaluator`.

plafer replied:
done

irakliyk commented on prover/src/constraints/evaluator/default.rs:
```diff
@@ -339,3 +405,191 @@ where
+// LAGRANGE CONSTRAINTS BATCH EVALUATOR
+// ================================================================================================
+
+/// Contains a specific strategy for evaluating the Lagrange kernel boundary and transition
+/// constraints where the divisors' evaluation is batched.
+///
+/// Specifically, [`batch_inversion`] is used to reduce the number of divisions performed.
+struct LagrangeConstraintsBatchEvaluator<E: FieldElement> {
+    lagrange_kernel_boundary_constraint: Option<LagrangeKernelBoundaryConstraint<E>>,
+    lagrange_kernel_transition_constraints: Option<LagrangeKernelTransitionConstraints<E>>,
+    aux_rand_elements: AuxTraceRandElements<E>,
+}
```
I would move this to `src/constraints/evaluator/lagrange.rs` module. Also, as mentioned in the previous comment, if we make the whole filed for `LagrangeConstraintsBatchEvaluator` optional, we won't need to make the internal fields optional.

Lastly, `aux_rand_elements` are the $r$-values, right? If so, I'd change this filed to `rand_elements: Vec<E>` as these values will be different from `AuxTraceRandElements`.

irakliyk commented on prover/src/constraints/evaluator/default.rs:
```diff
@@ -257,6 +288,41 @@ where
+        // LDE domain
+        let lde_shift = domain.ce_to_lde_blowup().trailing_zeros();
+
+        let mut lagrange_kernel_column_frames =
+            vec![LagrangeKernelEvaluationFrame::<E>::new_empty(); domain.ce_domain_size()];
+
+        for (step, frame) in lagrange_kernel_column_frames.iter_mut().enumerate() {
+            trace.read_lagrange_kernel_frame_into(
+                step << lde_shift,
+                lagrange_kernel_aux_column_idx,
+                frame,
+            );
+        }
```
If I'm understanding this correctly, this would store a frame per row in the constraint evaluation domain, right? If so, this is probably too expensive from memory standpoint. For example, for a trace with $2^20$ columns, this would be equivalent to having extra 20 columns in the auxiliary trace, and this would significantly increase RAM requirements (e.g., in some cases doubling them).

This interferes a bit with batch inversion of denominators approach (would have been really nice if we could pre-compute them for all constraints in a single vector) - but I think we can still make it work if we change the methodology slightly: we still evaluate one constraint at a time, but instead of working with evaluation frames, we work directly with the trace (e.g., read elements needed only for the current constraint). Maybe there are other ways of doing this too.

irakliyk reviewed (changes requested):
Thank you! I left a few more comments inline. The main ones are about segregating Lagrange kernel constraints from the rest of the structure.

Also, could you rebase this from the latest `main` - handling of `no_std` changed there slight.

Regarding benchmarks: what exactly are these measuring (i.e., is it just Lagrange column constraint evaluation - or something else)? If this is just for evaluating constraints, we'll need to dig deeper into how to optimize this as it seems like the impact on prover performance would be quite bad (i.e., it will slow down proof generation for something like Miden VM by 10% - 20% where generating proof for $2^{16}$ trace takes about about 2.5 seconds on a single core).

plafer commented:
> Also, could you rebase this from the latest main - handling of no_std changed there slight.

I merged `main` in, but it seems like it made the diff against `next` bigger. Is this intended?

> Regarding benchmarks: what exactly are these measuring (i.e., is it just Lagrange column constraint evaluation - or something else)? If this is just for evaluating constraints, we'll need to dig deeper into how to optimize this as it seems like the impact on prover performance would be quite bad

It is [measuring the entire](https://github.com/plafer/winterfell/blob/40a3b06dd51f62cfbf5a3f536e91480981202a86/prover/benches/lagrange_kernel.rs#L33) `Prover::prove()` function - not just the Lagrange constraint evaluation. I did this because the prover API is likely more stable than any internal Lagrange kernel API for constraint evaluation - but I can change it to just benchmark the Lagrange constraint evaluation if that is more useful. I personally used it more as a means of knowing if any given commit did indeed improve general performance, as opposed to knowing the absolute Lagrange kernel constraint evaluation time.

plafer commented on air/src/air/coefficients.rs:
```diff
@@ -66,8 +66,16 @@ impl<E: FieldElement> Default for AuxTraceRandElements<E> {
+pub struct LagrangeConstraintsCompositionCoefficients<E: FieldElement> {
+    pub transition: Vec<E>,
+    pub boundary: Option<E>,
+}
```
done

irakliyk reviewed (changes requested):
Thank you! Looks good! This is not a full review - but I did leave some more comments inline.

plafer commented on air/src/air/divisor.rs:
```diff
@@ -54,10 +54,6 @@ impl<B: StarkField> ConstraintDivisor<B> {
-        assert!(
-            num_exemptions > 0,
-            "invalid number of transition exemptions: must be greater than zero"
-        );
         let exemptions = (trace_length - num_exemptions..trace_length)
             .map(|step| get_trace_domain_value_at::<B>(trace_length, step))
             .collect();
```
This would re-introduce some Lagrange kernel concepts into the rest of the API (something we've been trying to avoid elsewhere). I also don't think this `assert!` is necessary, since `AirContext`'s constructor already checks it (e.g. for [main trace](https://github.com/plafer/winterfell/blob/9ea6c2df597e1fc110a93b6f94058de18a840b96/air/src/air/context.rs#L103)).

plafer commented on air/src/air/lagrange/boundary.rs:
```diff
@@ -0,0 +1,60 @@
+    E: FieldElement,
+{
+    /// Creates a new Lagrange kernel boundary constraint from the specified single assertion.
+    pub fn new(composition_coefficient: E, lagrange_kernel_rand_elements: &[E]) -> Self {
```
fixed

plafer commented on air/src/air/lagrange/boundary.rs:
```diff
@@ -0,0 +1,60 @@
+        numerator / denominator
+    }
+
+    /// Returns the evaluation of the boundary constraint numerator.
+    ///
+    /// `frame` is the evaluation frame of the Lagrange kernel column `c`, starting at `c(x)` for some `x`
+    pub fn evaluate_numerator_at(&self, frame: &LagrangeKernelEvaluationFrame<E>) -> E {
```
Updated

plafer commented on air/src/air/mod.rs:
```diff
@@ -357,13 +382,13 @@ pub trait Air: Send + Sync {
     fn get_boundary_constraints<E: FieldElement<BaseField = Self::BaseField>>(
         &self,
         aux_rand_elements: &AuxTraceRandElements<E>,
-        composition_coefficients: &[E],
+        boundary_composition_coefficients: &[E],
     ) -> BoundaryConstraints<E> {
         BoundaryConstraints::new(
             self.context(),
             self.get_assertions(),
             self.get_aux_assertions(aux_rand_elements),
-            composition_coefficients,
+            boundary_composition_coefficients,
         )
     }
```
fixed

plafer commented on air/src/proof/context.rs:
```diff
@@ -215,9 +215,9 @@ mod tests {
-            [aux_rands],
+            main_width as usize,
+            [aux_width as usize],
+            [aux_rands as usize],
```
Fixed. I would have assumed `clippy` would catch that

plafer commented on prover/src/constraints/evaluator/boundary.rs:
```diff
@@ -34,10 +34,11 @@ impl<E: FieldElement> BoundaryConstraints<E> {
     pub fn new<A: Air<BaseField = E::BaseField>>(
         air: &A,
         aux_rand_elements: &AuxTraceRandElements<E>,
-        composition_coefficients: &[E],
+        boundary_composition_coefficients: &[E],
     ) -> Self {
         // get constraints from the AIR instance
-        let source = air.get_boundary_constraints(aux_rand_elements, composition_coefficients);
+        let source =
+            air.get_boundary_constraints(aux_rand_elements, boundary_composition_coefficients);
```
removed prefix

plafer commented on verifier/src/lib.rs:
```diff
@@ -200,15 +200,17 @@ where
     let ood_trace_frame = channel.read_ood_trace_frame();
     let ood_main_trace_frame = ood_trace_frame.main_frame();
     let ood_aux_trace_frame = ood_trace_frame.aux_frame();
+    let ood_lagrange_kernel_column_frame = ood_trace_frame.lagrange_kernel_frame();
```
Renamed to `ood_lagrange_kernel_frame` (which is what I assume you meant?)

plafer commented:
Addressed comments, and further optimized constraint evaluation. 

## Timing
For a trace length of $2^{20}$, timings are:

**trace column building:** 84ms
**precompute transition divisors time:** 11ms
**invert transition divisors time:** 34ms
**invert boundary divisors time:** 19ms
**evaluate and merge constraints time:** 387ms

**TOTAL lagrange constraints eval time:** 452ms

# Benchmarks
Benchmarks are also now significantly faster:

- trace length $2^{16}$: [163.01 ms 165.96 ms 168.60 ms]
- trace length $2^{20}$: [3.0080 s 3.0399 s 3.0690 s]

irakliyk commented on prover/src/constraints/evaluator/lagrange.rs:
```diff
@@ -0,0 +1,289 @@
+            // The number of divisor evaluations is
+            // `ce_domain_size + ce_domain_size/2 + ce_domain_size/4 + ... + ce_domain_size/(log(ce_domain_size)-1)`,
+            // which is slightly smaller than `ce_domain_size * 2`
+            let mut divisor_evals: Vec<E> = Vec::with_capacity(domain.ce_domain_size() * 2);
```
Probably won't change things much, but we frequently use `uninit_vector()` function to skip memory initialization when the whole vector will be overwritten.

plafer replied:
`Vec::with_capacity()` doesn't initialize the memory, as discussed in the [Guarantees section of the docs](https://doc.rust-lang.org/std/vec/struct.Vec.html#guarantees) (see the diagram). So unless I'm missing something, the following 2 approaches generate the same code:
- construct vector with `Vec::with_capacity()`, and `Vec::push()` elements one by one
- construct vec with `uninit_vector()`, and write elements one by one with `v[i] = ...;`

Since `Vec::with_capacity()` is canonical, and doesn't involve exposing uninitialize memory, I think it should be preferred over `uninit_vector()`. Unless I'm missing something?

irakliyk replied:
In the testing I did a while back (so, things may have changed since then) using un-initialized vector and then setting values via `v[I] = x` was noticeably faster than doing `Vec::with_capacity() ` and then using `.push()` method. Was not a huge difference, but iirc it was somewhere between 10% and 20%.

plafer replied:
Oh yes you're right, I just benchmarked it, and indeed I using `uninit_vector()` was 40% faster! I will change the code to use it.

plafer replied:
Actually it makes the code much more complex, since

1. we need to compute the exact size of the vector, which is

```rust
let mut divisor_evals = unsafe {
    uninit_vector(
        domain.ce_domain_size() * 2
            - (domain.ce_domain_size()
                / 2_usize.pow(num_lagrange_transition_constraints as u32 - 1)),
    )
};
```

2. the index to which we need to write to is a complex function of `trans_constraint_idx` and `step`, so we'd be better off introducing an `idx` variable which we increment after each write (which makes the code that much harder to read).

I tried it, ran the benchmarks, and it didn't change the performance. So I would just leave it as is.

irakliyk commented on prover/src/constraints/evaluator/lagrange.rs:
```diff
@@ -0,0 +1,289 @@
+        E: FieldElement<BaseField = A::BaseField>,
+    {
+        Self {
+            lagrange_kernel_constraints: LagrangeKernelConstraints::new(
+                air.context(),
+                lagrange_composition_coefficients,
+                &lagrange_kernel_rand_elements,
+            ),
```
Why not use `air.get_lagrange_kernel_constraints()` here instead of `LagrangeKernelConstraints::new()`?

plafer replied:
Changed to `air.get_lagrange_kernel_constraints()`; however since it returns an `Option<>`, we have to use an `expect()`, which arguably is not as nice?

irakliyk commented on prover/src/constraints/evaluator/default.rs:
```diff
@@ -107,8 +110,28 @@ where
+        let combined_evaluations = {
+            let main_and_aux_evaluations = evaluation_table.combine();
+
+            match self.evaluate_lagrange_kernel_constraints(trace, domain) {
+                Some(lagrange_kernel_combined_evals) => {
+                    // if present, linearly combine the Lagrange kernel evaluations too
+                    debug_assert_eq!(
+                        main_and_aux_evaluations.len(),
+                        lagrange_kernel_combined_evals.len()
+                    );
+                    main_and_aux_evaluations
+                        .into_iter()
+                        .zip(lagrange_kernel_combined_evals)
+                        .map(|(eval_1, eval_2)| eval_1 + eval_2)
+                        .collect()
+                }
+                None => main_and_aux_evaluations,
+            }
```
This is probably fine for now, but we may be able to optimize this a bit more in the future. Specifically: right now we first compute the entire vector with Lagrange constraint evaluations, and then add the values to the main/aux evaluations. By refactoring this a bit (and passing `main_and_aux_evaluations` into the Lagrange constraint evaluator we could avoid the second loop (i.e., we'd treat `main_and_aux_evaluations` as an accumulator to which Lagrange constraint evaluations get added immediately as they are computed for each row).

Let's add a `TODO` here and also create an issue for this.

plafer replied:
I implemented it and got the exact same performance (within noise bounds):

```
trace column building: 84ms
precompute trans divisors time: 11ms
invert trans divisors time: 34ms
invert boundary divisors time: 21ms
evaluate and merge constraints time: 404ms

TOTAL lagrange constraints eval time: 472ms
```

Since the first one is more readable and in a functional style, I would suggest we revert that [commit](https://github.com/facebook/winterfell/pull/247/commits/7c74d894a835f3734da38cb2cee02723605cb7d9). What do you think?

irakliyk replied:
I don't have a strong preference here - so, fine with either of the approaches.

irakliyk commented on air/src/air/lagrange/mod.rs:
```diff
@@ -0,0 +1,37 @@
+
+use crate::{AirContext, LagrangeConstraintsCompositionCoefficients};
+
+/// Represents the Lagrange kernel transition and boundary constraints.
+pub struct LagrangeKernelConstraints<E: FieldElement> {
+    pub transition: LagrangeKernelTransitionConstraints<E>,
+    pub boundary: LagrangeKernelBoundaryConstraint<E>,
+}
```
I wonder if the Lagrange column index should be a part of this struct. Would slightly simplify the code in a few places.

plafer replied:
done

irakliyk commented on verifier/src/evaluator.rs:
```diff
@@ -42,13 +46,17 @@ pub fn evaluate_constraints<A: Air, E: FieldElement<BaseField = A::BaseField>>(
-            &aux_rand_elements,
-            &mut t_evaluations2,
-        );
+        // `t_evaluations_2` would be empty here if there is only 1 auxiliary column, and that
+        // column is the Lagrange kernel column
+        if !t_evaluations2.is_empty() {
```
Should this be an error as a trace with just one Lagrange column is not really useful?

plafer replied:
Oops, forgot to remove that one. Fixed

plafer commented on air/src/air/lagrange/boundary.rs:
```diff
@@ -0,0 +1,60 @@
+        }
+    }
+
+    /// Returns the evaluation of the boundary constraint at `x`.
+    ///
+    /// `frame` is the evaluation frame of the Lagrange kernel column `c`, starting at `c(x)`
+    pub fn evaluate_at(&self, x: E, frame: &LagrangeKernelEvaluationFrame<E>) -> E {
```
done

irakliyk reviewed (changes requested):
Thank you! Looks good! I left some more comments inline - but these are mostly nits.

One thing I'm still trying to think through is whether we should change anything about how we handle OOD frame structs. But I think this is the last potential change (and I'm not sure if anything should be changed at all on this front).

@Al-Kindi-0 - could you give this another review?

irakliyk commented on prover/src/trace/poly_table.rs:
```diff
@@ -64,11 +72,27 @@ impl<E: FieldElement> TracePolyTable<E> {
+    /// domain. Additionally, if the Lagrange kernel auxiliary column is present, we also evaluate
+    /// that column over the points: z, z * g, z * g^2, z * g^4, ..., z * g^(2^(v-1)), where v =
+    /// log(trace_len).
+    pub fn get_ood_frame(&self, z: E) -> OodFrameTraceStates<E> {
+        let log_trace_len = self.poly_size().ilog2();
+        let g = E::from(E::BaseField::get_root_of_unity(log_trace_len));
+        let current_frame = self.evaluate_at(z);
+        let next_frame = self.evaluate_at(z * g);
+
+        let lagrange_kernel_frame = self.lagrange_kernel_column_idx.map(|col_idx| {
+            let lagrange_kernel_col_poly = self.aux_segment_polys[0].get_column(col_idx);
+
+            LagrangeKernelEvaluationFrame::from_lagrange_kernel_column_poly(
+                lagrange_kernel_col_poly,
+                z,
+            )
+        });
+
+        OodFrameTraceStates::new(current_frame, next_frame, lagrange_kernel_frame)
```
A few more thoughts about OOD frame structs.

It seems to me that we have two structs which contain basically identical data: `OodFrameTraceStates` (in the `air`  crate, but used primarily in the prover) and `TraceOodFrame` (used in the verifier). I wonder if we can unify these. For example, we could:

1. Replace the use of `TraceOodFrame` with `OodFrameTraceStates`. We might need to add some more methods to `OodFrameTraceStates` to do this - but I am not seeing any fundamental reasons why we can't do that.
2. Rename `OodFrameTraceStates` into `TraceOodFrame` (I like this name a little better).
3. Update the filed names as I mentioned in one of the previous comments. The struct can then look like this:

```Rust
pub struct TraceOodFrame <E: FieldElement> {
    current_row: Vec<E>,
    next_row: Vec<E>,
    lagrange_kernel_frame: Option<LagrangeKernelEvaluationFrame<E>>,
}
```

Then, `TracePolyTable::get_ood_frame()` method would return this struct and it would be passed to the prover channel.

One the verifier side, I think `OodFrame::parse()` could return this struct as well. The signature could look something like this:
```Rust
pub fn parse<E: FieldElement>(
    self,
    main_trace_width: usize,
    aux_trace_width: usize,
    num_evaluations: usize,
) -> Result<(TraceOodFrame<E>, Vec<E>), DeserializationError>
```

This should also nicely encapsulate the "interleaving" logic between current and next rows which is currently split between `OodFrame` and `TraceOodFrame`.

plafer replied:
I am all for this. Should we create a new issue though? This suggestion is pretty self-contained, and this PR is already very big.

Al-Kindi-0 reviewed (commented):
Looks great, thank you!
Left some small comments inline.
There is one thing which should conclude this PR and that is updating the portion related to the DEEP composition polynomial. I am not sure if this should go into this PR (as this one is already big enough and the remaining changes are quite substantial). I think a good improvement to discuss that will be useful for the remaining changes is to see if we can do away with using the coefficient form of the trace column polynomials. This will have some performance advantages, I believe, as well as make some of the changes required to accommodate the Lagrange kernel easy.

plafer commented on air/src/air/mod.rs:
```diff
@@ -285,6 +292,23 @@ pub trait Air: Send + Sync {
+        lagrange_composition_coefficients: LagrangeConstraintsCompositionCoefficients<E>,
+        lagrange_kernel_rand_elements: &[E],
+    ) -> Option<LagrangeKernelConstraints<E>> {
+        self.context().lagrange_kernel_aux_column_idx().map(|col_idx| {
+            LagrangeKernelConstraints::new(
+                self.context(),
+                lagrange_composition_coefficients,
+                lagrange_kernel_rand_elements,
+                col_idx,
+            )
+        })
```
Agreed. I actually just fully removed it, and removed the `assert!()` statement that checks that the length of `lagrange_composition_coefficients` is `log(trace_len)`. I find it cleaner to just have `LagrangeKernelConstraints` assume that the arguments are well-formed (and leaving that concern to other parts of the codebase).

plafer commented on air/src/air/mod.rs:
```diff
@@ -501,9 +525,28 @@ pub trait Air: Send + Sync {
             b_coefficients.push(public_coin.draw()?);
         }
 
+        let lagrange = if self.context().has_lagrange_kernel_aux_column() {
+            let mut lagrange_kernel_t_coefficients = Vec::new();
+            if self.context().has_lagrange_kernel_aux_column() {
+                for _ in 0..self.context().trace_len().ilog2() {
+                    lagrange_kernel_t_coefficients.push(public_coin.draw()?);
+                }
+            }
```
Oops, it wasn't; that was just an editing error.

plafer commented on prover/src/constraints/evaluator/lagrange.rs:
```diff
@@ -0,0 +1,289 @@
+            batch_inversion(&divisor_evals)
+        };
+
+        let slice_indices_precomputes = {
+            let num_indices = num_lagrange_transition_constraints + 1;
+            let mut slice_indices_precomputes = Vec::with_capacity(num_indices);
+
+            slice_indices_precomputes.push(0);
+
+            let mut current_slice_len = domain.ce_domain_size();
+            for i in 1..num_indices {
+                let next_precompute_index = slice_indices_precomputes[i - 1] + current_slice_len;
+                slice_indices_precomputes.push(next_precompute_index);
+
+                current_slice_len /= 2;
+            }
+
+            slice_indices_precomputes
+        };
```
Right. `get_inverse_divisor_eval()` is called for each index of the constraint evaluation domain, and computing this slice index requires 1 or more divisions (unless there's a closed form expression that I'm not aware of).

So without this precomputation, we'd be back to running a division at every index.

plafer commented on air/src/proof/ood_frame.rs:
```diff
@@ -149,13 +194,68 @@ impl Deserializable for OodFrame {
+/// the Air contains a Lagrange kernel auxiliary column, then that column interpolated polynomial
+/// will be evaluated at `z`, `gz`, `g^2 z`, ... `g^(2^(v-1)) z`, where `v == log(trace_len)`, and
+/// stored in `lagrange_kernel_frame`.
+pub struct OodFrameTraceStates<E: FieldElement> {
+    current_frame: Vec<E>,
+    next_frame: Vec<E>,
+    lagrange_kernel_frame: Option<LagrangeKernelEvaluationFrame<E>>,
```
Agreed with `current_row` and `next_row` (updated). However I prefer `lagrange_kernel_frame`, as we already use "Lagrange kernel frame" elsewhere in the codebase to refer to these "Lagrange kernel values" - so this way it's more consistent IMO.

plafer commented on air/src/air/context.rs:
```diff
@@ -100,10 +103,12 @@ impl<B: StarkField> AirContext<B> {
         assert!(num_main_assertions > 0, "at least one assertion must be specified");
 
         if trace_info.is_multi_segment() {
+            // If the only auxiliary column is the Lagrange kernel one, then we don't require any
+            // other boundary/transition constraints
```
removed

plafer commented on prover/src/trace/trace_lde/default/mod.rs:
```diff
@@ -174,6 +175,28 @@ where
+
+        let frame_length = self.trace_info.length().ilog2() as usize + 1;
+        for i in 0..frame_length - 1 {
+            let shift = self.blowup() * 2_u32.pow(i as u32) as usize;
```
Switched to `(1 << i)` - this is arguably more readable due to the removal of both `as _`.

The benchmarks indeed didn't improve though

irakliyk reviewed (approved):
All looks good! Thank you! As the next steps, we should do the following:

1. Make a small refactoring mentioned [here](https://github.com/facebook/winterfell/pull/247#discussion_r1537123827) and also remove the incomplete machinery related to supporting multiple auxiliary trace segments.
2. Complete the ability to run the full GKR protocol after the main trace commitment is computed (discussed in [#257](https://github.com/facebook/winterfell/issues/257). This may require refactoring how our prover/verifier channels work.
3. Implement proper handling of DEEP composition for the Lagrange constraints.
